### PR TITLE
a11y: Unify screen reader in visual editor

### LINKED ISSUE #2079: [Accessibility][Keyboard -Create Node]: Status message for the node creation is not announced by the narrator.
**User Experience** 
The AT user might gets confused as no confirmation message is announced for the node creation. If the status messages are not narrated by the narrator then the screen reader dependent user will not be able to know the confirmation of the actions performed by him.

**Test Environment**
Browser: Microsoft Edge Canary Version: 82.0.425.0 (Official build)
OS build: 2004 (19564.1005)
Screen Reader: Narrator
URL: [Bot Framework Composer](http://localhost:3000/dialogs/Main)
[**Prerequisite Here**](https://github.com/microsoft/BotFramework-Composer)

**Repro Steps:** 
1. Open the above mentioned URL in EDGE and set the zoom at 400%.
2. Select 'Design Flow' from left navigation.
3. Now, navigate to the flow chart.
4. Navigate to "+" button and activate it.
5. Create a node by selecting any of the available.
6. Observe the issue.

**Actual Result:**
Screen reader doesn’t provide any confirmation on creating 'New node'. New node gets added but screen reader doesn't provide any confirmation.

**Expected:**
The narrator should announce any status message such as it should announce as "New node is created".

**Note:**
Same issue observed throughout the "Design Flow" page such as copy, cut, paste, node creation, Delete, New trigger creation, new dialog creation etc.

**MAS Impacted:** [MAS4.1.3](https://microsoft.sharepoint.com/:w:/r/teams/msenable/_layouts/15/Doc.aspx?sourcedoc=%7B684573A7-B089-4131-9B39-0009054125D3%7D&file=MAS%204.1.3%20%E2%80%93%20Status%20Messages.docx&action=default&mobileredirect=true&cid=6a538745-b402-41f3-85ea-9f67151fa4c9)

**Attachment:**
[MAS4.1.3_Status messages are not announced.zip](https://github.com/microsoft/BotFramework-Composer/files/4249592/MAS4.1.3_Status.messages.are.not.announced.zip)

### PR DESCRIPTION
**Pull request by alanlong9278**
## Description

Unify screen reader in visual editor.
```
  EventFocused = 'Event focused',
  ActionFocused = 'Action focused',
  DialogOpened = 'Dialog opened',
  ActionDeleted = 'Action deleted',
  ActionsDeleted = 'Actions deleted',
  ActionCreated = 'Action created',
  ActionsCreated = 'Actions created',
  EventCreated = 'Event created',
  ActionsCopied = 'Actions copied',
  ActionsCut = 'Actions cut',
  ActionsMoved = 'Actions moved',
  ActionUndo = 'Undo',
  ActionRedo = ' Redo',
```

## Task Item

fix #2079

### CODE REVIEW

File: Composer/packages/extensions/visual-designer/src/constants/ScreenReaderMessage.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  ActionsDeleted = 'Actions deleted',
+  ActionCreated = 'Action created',
+  ActionsCreated = 'Actions created',
+  EventCreated = 'Event created',
+  ActionsCopied = 'Actions copied',
+  ActionsCut = 'Actions cut',
+  ActionsMoved = 'Actions moved',
+  ActionUndo = 'Action undo',
+  ActionRedo = 'Action redo',
+}
```
Comment (beyackle) [inline Composer/packages/extensions/visual-designer/src/constants/ScreenReaderMessage.ts:18]: I think these could just be "undo" and "redo".

File: Composer/packages/extensions/visual-designer/src/editors/ObiEditor.tsx [hunk 2]
```diff
@@ -256,13 +273,20 @@ export const ObiEditor: FC<ObiEditorProps> = ({
           if (!e.target || e.target === focusedEvent) return;
           const dialog = appendNodesAfter(data, e.target, e.actions);
           onChange(dialog);
+          announce(ScreenReaderMessage.ActionsCreated);
         };
         break;
       case NodeEventTypes.Undo:
-        handler = undo;
+        handler = () => {
+          undo && undo();
```
Comment (beyackle) [inline Composer/packages/extensions/visual-designer/src/editors/ObiEditor.tsx:281]: `undo?.()`

File: Composer/packages/extensions/visual-designer/src/editors/ObiEditor.tsx [hunk 3]
```diff
@@ -256,13 +273,20 @@ export const ObiEditor: FC<ObiEditorProps> = ({
...
+          announce(ScreenReaderMessage.ActionUndo);
+        };
         break;
       case NodeEventTypes.Redo:
-        handler = redo;
+        handler = () => {
+          redo && redo();
```
Comment (beyackle) [inline Composer/packages/extensions/visual-designer/src/editors/ObiEditor.tsx:287]: `redo?.()`

File: Composer/packages/extensions/visual-designer/src/constants/ScreenReaderMessage.ts (hunk 1 above)
Comment (alanlong9278) [reply Composer/packages/extensions/visual-designer/src/constants/ScreenReaderMessage.ts:18]: Sure.

**Comment by yeze322** [comment]
@alanlong9278 This is a good catch up of all scenarios invoke the ScreenReader in Visual Editor!
___
As a holistic work, I think it would be nice to have global shared handlers (hook) to manage SceenReader messages considering 3 points:
1. Not only Visual Editor but the **whole Composer** needs a consistent exp when triggering UI events that require to be announced. The module should be shared by Shell/Visual/Extensions.
2. We may need to define **more semantic event apis** whose name describes when and why it triggers an announcement. For example, `userCreatedDialog`, `userPressedKeyDown`.
**Motivation**: we need an overview of all screen reader events and messages instead of just invoking `announce` with a message everywhere. Invoker should have a solid reason to invoke that event.
3. We should control the life cycle of screen reader events. Screen Reader can only read 1 message at a time, if a new message comes in, the previous one should be canceled.
___
As an example, we can put the shared hook under the `@bfc/extension` package
```ts
const useAnnounceApi = () => {
   const { announce } = shellApi;
   return {
     userCreatesDialog: () => announce(Hints.DialogCreated),
     userReachesEndOfRow: (keyname: string) => announce(ReachEndOfRowLabel),
   }
}
```
It imports a group of constant messages
```ts
const Hints = {
   DialogCreated: formatMessage('Dialog created.'),
   ReachEndOfRowLabel: formatMessage('press Enter to add this item or Tab to move to the next interactive element')
}
```

**Comment by alanlong9278** [comment]
> @alanlong9278 This is a good catch up of all scenarios invoke the ScreenReader in Visual Editor!
> 
> As a holistic work, I think it would be nice to have unified global handlers (hook) to manage SceenReader messages considering 3 points:
> 
> 1. Not only Visual Editor but the **whole Composer** needs a consistent exp when triggering UI events that require to be announced. The module should be shared by Shell/Visual/Extensions.
> 2. We may need to define **more semantic event apis** whose name describes when and why it triggers an announcement. For example, `userCreatedDialog`, `userPressedKeyDown`.
>    **Motivation**:
>    
>    1. avoid exposing low level `announce` to invokers
>    2. we need an overview of all screen reader events and messages instead of just invoking `announce` with a message everywhere. Invoker should have a solid reason to invoke that event.
> 3. We should control the life cycle of screen reader events. Screen Reader can only read 1 message at a time, if a new message comes in, the previous one should be canceled.
> 
> As an example, we can put the shared hook under the `@bfc/extension` package
> 
> ```ts
> const useAnnouceApi = () => {
>   const { announce } = shellApi;
>   return {
>     announce,
>     AnnounceEvents: Hints,
>   }
> }
> // or a more semantic solution
> const useAnnounceApi = () => {
>    const { announce } = shellApi;
>    return {
>      userCreatesDialog: () => announce(Hints.DialogCreated),
>      userReachesEndOfRow: (keyname: string) => announce(ReachEndOfRowLabel),
>    }
> }
> ```
> 
> It imports a group of constant messages
> 
> ```ts
> const Hints = {
>    DialogCreated: formatMessage('Dialog created.'),
>    ReachEndOfRowLabel: formatMessage('press Enter to add this item or Tab to move to the next interactive element')
> }
> ```
> 
> In my opinion, when we are building the screen reader feature, we are actually building a notification feature. It's not quite different from an UI notification bubble, merely it's read by voice.

Thanks for advicing. @yeze322 . I will refactor the screen reader feature into composer after accessiblity issue fixed.

**Comment by cwhitten** [comment]
> @alanlong9278 This is a good catch up of all scenarios invoke the ScreenReader in Visual Editor!
> 
> As a holistic work, I think it would be nice to have unified global handlers (hook) to manage SceenReader messages considering 3 points:
> 
> 1. Not only Visual Editor but the **whole Composer** needs a consistent exp when triggering UI events that require to be announced. The module should be shared by Shell/Visual/Extensions.
> 2. We may need to define **more semantic event apis** whose name describes when and why it triggers an announcement. For example, `userCreatedDialog`, `userPressedKeyDown`.
>    **Motivation**:
>    
>    1. avoid exposing low level `announce` to invokers
>    2. we need an overview of all screen reader events and messages instead of just invoking `announce` with a message everywhere. Invoker should have a solid reason to invoke that event.
> 3. We should control the life cycle of screen reader events. Screen Reader can only read 1 message at a time, if a new message comes in, the previous one should be canceled.
> 
> As an example, we can put the shared hook under the `@bfc/extension` package
> 
> ```ts
> const useAnnouceApi = () => {
>   const { announce } = shellApi;
>   return {
>     announce,
>     AnnounceEvents: Hints,
>   }
> }
> // or a more semantic solution
> const useAnnounceApi = () => {
>    const { announce } = shellApi;
>    return {
>      userCreatesDialog: () => announce(Hints.DialogCreated),
>      userReachesEndOfRow: (keyname: string) => announce(ReachEndOfRowLabel),
>    }
> }
> ```
> 
> It imports a group of constant messages
> 
> ```ts
> const Hints = {
>    DialogCreated: formatMessage('Dialog created.'),
>    ReachEndOfRowLabel: formatMessage('press Enter to add this item or Tab to move to the next interactive element')
> }
> ```
> 
> In my opinion, when we are building the screen reader feature, we are actually building a notification feature. It's not quite different from an UI notification bubble, merely it's read by voice.

Hey @yeze322  this isn't exactly how we want the screenreader messages to behave. For example, we _dont_ necessarily want to cancel a message intended to be announced if a new one comes in. There are aria standards built into the web platform to help us with this. Let's follow-up on your requests here before we execute on this.

**Comment by beyackle** [comment]
I like the idea of having a library of consistent announcements (especially if we get some help from design on this so we know exactly what events ought to trigger an announcement), but I don't think we need to control the lifecycle of announcement events. Let the browser do that for us; it's what the aria-live region is there for, and that's what the `announce` thing on `useShell` hooks into.

**Comment by yeze322** [comment]
@cwhitten @benbrown thanks! I misunderstood how ScreenReader works

**Pull request merged**